### Final Project Requirements/notes: https://docs.google.com/document/d/1mwYbYJHkB7kpx4tNflKh54jN9_oOscw3p4k5fsmn3bc/edit

### Link with all Data: https://www.sec.gov/dera/data/financial-statement-and-notes-data-set.html
- using NUM file only for now (data set of all numeric XBRL facts presented on the primary financial statements)

In [1]:
import pandas as pd
q318numbers = pd.read_table('2018q3_notes/num.tsv', sep='\t')

In [2]:
q318numbers

adsh  \
0        0001477932-18-003294   
1        0001477932-18-003294   
2        0001477932-18-003294   
3        0001477932-18-003294   
4        0001477932-18-003294   
5        0001477932-18-003294   
6        0001477932-18-003294   
7        0001477932-18-003294   
8        0001477932-18-003294   
9        0001477932-18-003294   
10       0001477932-18-003294   
11       0001477932-18-003294   
12       0001477932-18-003294   
13       0001477932-18-003294   
14       0001477932-18-003294   
15       0001477932-18-003294   
16       0001477932-18-003294   
17       0001477932-18-003294   
18       0001477932-18-003294   
19       0001477932-18-003294   
20       0001477932-18-003294   
21       0001477932-18-003294   
22       0001477932-18-003294   
23       0001477932-18-003294   
24       0001477932-18-003294   
25       0001477932-18-003294   
26       0001477932-18-003294   
27       0001477932-18-003294   
28       0001477932-18-003294   
29       0001477932-18-003294   
...                       ...   
5372789  0001437749-18-017549   
5372790  0001437749-18-017549   
5372791  0001437749-18-017549   
5372792  0001437749-18-017549   
5372793  0001437749-18-017549   
5372794  0001437749-18-017549   
5372795  0001437749-18-017549   
5372796  0001437749-18-017549   
5372797  0001437749-18-017549   
5372798  0001437749-18-017549   
5372799  0001416265-18-000558   
5372800  0001416265-18-000558   
5372801  0001416265-18-000558   
5372802  0001416265-18-000558   
5372803  0001416265-18-000558   
5372804  0001416265-18-000558   
5372805  0001416265-18-000558   
5372806  0001416265-18-000558   
5372807  0001416265-18-000558   
5372808  0001416265-18-000558   
5372809  0001416265-18-000558   
5372810  0001416265-18-000558   
5372811  0001416265-18-000558   
5372812  0001416265-18-000558   
5372813  0001416265-18-000558   
5372814  0001416265-18-000558   
5372815  0001416265-18-000558   
5372816  0001416265-18-000558   
5372817  0001416265-18-000558   
5372818  0001416265-18-000558   

                                                       tag  \
0                AccountsNotesAndLoansReceivableNetCurrent   
1                AccountsNotesAndLoansReceivableNetCurrent   
2                AccountsNotesAndLoansReceivableNetCurrent   
3                     AccountsPayableRelatedPartiesCurrent   
4                     AccountsPayableRelatedPartiesCurrent   
5                     AccountsPayableRelatedPartiesCurrent   
6                             AccountsReceivableNetCurrent   
7                             AccountsReceivableNetCurrent   
8                             AccountsReceivableNetCurrent   
9        AccretionAmortizationOfDiscountsAndPremiumsInv...   
10       AccretionAmortizationOfDiscountsAndPremiumsInv...   
11                                                  Assets   
12                                                  Assets   
13                                                  Assets   
14                      AdditionalPaidInCapitalCommonStock   
15                      AdditionalPaidInCapitalCommonStock   
16                      AdditionalPaidInCapitalCommonStock   
17                       AmortizationOfDebtDiscountPremium   
18                       AmortizationOfDebtDiscountPremium   
19                       AmortizationOfDebtDiscountPremium   
20                       AmortizationOfDebtDiscountPremium   
21                               AssetsFairValueDisclosure   
22                               AssetsFairValueDisclosure   
23                               AssetsFairValueDisclosure   
24                               AssetsFairValueDisclosure   
25                               AssetsFairValueDisclosure   
26                               AssetsFairValueDisclosure   
27                               AssetsFairValueDisclosure   
28                               AssetsFairValueDisclosure   
29                               AssetsFairValueDisclosure   
...                                                    .

In [3]:
#get details on all columns
q318numbers.describe(include='all')

adsh  \
count                5372819   
unique                  6492   
top     0001370368-18-000041   
freq                   13799   
mean                     NaN   
std                      NaN   
min                      NaN   
25%                      NaN   
50%                      NaN   
75%                      NaN   
max                      NaN   

                                                      tag       version  \
count                                             5372819       5372819   
unique                                             193156          6357   
top     RevenueFromContractWithCustomerExcludingAssess...  us-gaap/2018   
freq                                                85042       2416424   
mean                                                  NaN           NaN   
std                                                   NaN           NaN   
min                                                   NaN           NaN   
25%                                                   NaN           NaN   
50%                                                   NaN           NaN   
75%                                                   NaN           NaN   
max                                                   NaN           NaN   

               ddate          qtrs      uom        dimh          iprx  \
count   5.372819e+06  5.372819e+06  5372819     5372819  5.372819e+06   
unique           NaN           NaN     2590      354096           NaN   
top              NaN           NaN      USD  0x00000000           NaN   
freq             NaN           NaN  4574527     2173578           NaN   
mean    2.017505e+07  1.046295e+00      NaN         NaN  8.248742e-03   
std     7.388262e+03  1.351834e+00      NaN         NaN  1.231226e-01   
min     1.968123e+07  0.000000e+00      NaN         NaN  0.000000e+00   
25%     2.017063e+07  0.000000e+00      NaN         NaN  0.000000e+00   
50%     2.018033e+07  1.000000e+00      NaN         NaN  0.000000e+00   
75%     2.018063e+07  2.000000e+00      NaN         NaN  0.000000e+00   
max     2.108073e+07  3.600000e+02      NaN         NaN  1.500000e+01   

               value                                           footnote  \
count   5.251128e+06                                              13335   
unique           NaN                                              12191   
top              NaN  Unrealized Appreciation (Depreciation) and Val...   
freq             NaN                                                 11   
mean    2.140916e+10                                                NaN   
std     3.559570e+12                                                NaN   
min    -1.534799e+13                                                NaN   
25%     2.900000e-01                                                NaN   
50%     1.300000e+06                                                NaN   
75%     3.680000e+07                                                NaN   
max     5.400011e+15                                                NaN   

             footlen          dimn             coreg          durp  \
count   5.372819e+06  5.372819e+06             78755  5.372819e+06   
unique           NaN           NaN              3906           NaN   
top              NaN           NaN  ConsolidatedBank           NaN   
freq             NaN           NaN               862           NaN   
mean    6.090959e-01  9.305514e-01               NaN  8.523703e-03   
std     2.241506e+01  9.880578e-01               NaN  3.355416e-02   
min     0.000000e+00  0.000000e+00               NaN -4.986305e-01   
25%     0.000000e+00  0.000000e+00               NaN  0.000000e+00   
50%     0.000000e+00  1.000000e+00               NaN  8.219957e-03   
75%     0.000000e+00  1.000000e+00               NaN  2.465802e-02   
max     1.790700e+04  9.000000e+00               NaN  4.986310e-01   

                datp          dcml  
count   5.372819e+06  5.372819e+06  
unique           NaN           NaN  
top     

### From "Financial Statement and Notes Data Sets" Readme:
These fields comprise a unique compound key:

1) **adsh - EDGAR accession number**: a unique identifier assigned automatically to an accepted submission by the EDGAR Filer System; The first set of numbers (0001193125) is the CIK of the entity submitting the filing. The next 2 numbers (18) represent the year. The last series of numbers represent a sequential count of submitted filings from that CIK. The count is usually, but not always, reset to 0 at the start of each calendar year.
- **TODO**: separate these numbers to identify a company or a financial filing, there were 6,492 individual filings

2) **tag** - tag used by the filer 
- **TODO**: may have to separate out first word from tag to identify broader groups such as revenue

3) **version** – if a standard tag, the taxonomy of origin, otherwise equal to adsh.

4) **ddate** - period end date

5) **qtrs** - duration in number of quarters

6) **uom** - unit of measure

7) **dimh** - 16-byte dimensional qualifier

8) **iprx** - a sequential integer used to distinguish otherwise identical facts

9) **coreg** - If specified, indicates a specific co-registrant, the parent company, or other entity (e.g., guarantor).  NULL indicates the consolidated entity.  Note that this value is a function of the dimension segments.

10) **durp** - The difference between the reported fact duration and the quarter duration (qtrs), expressed as a fraction of 1.  For example, a fact with duration of 120 days rounded to a 91-day quarter has a durp value of 29/91 = +0.3187.

11) **datp** - The difference between the reported fact date and the month-end rounded date (ddate), expressed as a fraction of 1.  For example, a fact reported for 29/Dec, with ddate rounded to 31/Dec, has a datp value of minus 2/31 = -0.0645.
 
12) **dcml** - The value of the fact "decimals" attribute, with INF represented by 32767.

In [4]:
q318numbers.isnull().sum()

adsh              0
tag               0
version           0
ddate             0
qtrs              0
uom               0
dimh              0
iprx              0
value        121691
footnote    5359484
footlen           0
dimn              0
coreg       5294064
durp              0
datp              0
dcml              0
dtype: int64

#### A lot of null values for footnotes and coregistrants (majority of rows); will remove these columns for now

In [5]:
q318numbers = q318numbers.drop(columns=['footnote','coreg'])

In [6]:
q318numbers.isnull().sum() 

adsh            0
tag             0
version         0
ddate           0
qtrs            0
uom             0
dimh            0
iprx            0
value      121691
footlen         0
dimn            0
durp            0
datp            0
dcml            0
dtype: int64

#### Didn't get rid of blank Value fields, so just get rid of those since there will still be enough data points to analyze

In [7]:
q318numbers = q318numbers.dropna()

In [9]:
q318numbers.describe(include='all') #still have 5,251,128 data points for Q3

adsh  \
count                5251128   
unique                  6492   
top     0001370368-18-000041   
freq                   13799   
mean                     NaN   
std                      NaN   
min                      NaN   
25%                      NaN   
50%                      NaN   
75%                      NaN   
max                      NaN   

                                                      tag       version  \
count                                             5251128       5251128   
unique                                             192457          6354   
top     RevenueFromContractWithCustomerExcludingAssess...  us-gaap/2018   
freq                                                84561       2342159   
mean                                                  NaN           NaN   
std                                                   NaN           NaN   
min                                                   NaN           NaN   
25%                                                   NaN           NaN   
50%                                                   NaN           NaN   
75%                                                   NaN           NaN   
max                                                   NaN           NaN   

               ddate          qtrs      uom        dimh          iprx  \
count   5.251128e+06  5.251128e+06  5251128     5251128  5.251128e+06   
unique           NaN           NaN     2589      351622           NaN   
top              NaN           NaN      USD  0x00000000           NaN   
freq             NaN           NaN  4463505     2122035           NaN   
mean    2.017507e+07  1.037257e+00      NaN         NaN  8.424857e-03   
std     7.369952e+03  1.345961e+00      NaN         NaN  1.244706e-01   
min     1.968123e+07  0.000000e+00      NaN         NaN  0.000000e+00   
25%     2.017063e+07  0.000000e+00      NaN         NaN  0.000000e+00   
50%     2.018033e+07  1.000000e+00      NaN         NaN  0.000000e+00   
75%     2.018063e+07  2.000000e+00      NaN         NaN  0.000000e+00   
max     2.108073e+07  3.600000e+02      NaN         NaN  1.500000e+01   

               value       footlen          dimn          durp          datp  \
count   5.251128e+06  5.251128e+06  5.251128e+06  5.251128e+06  5.251128e+06   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean    2.140916e+10  6.034273e-01  9.336223e-01  8.484013e-03 -2.223713e-02   
std     3.559570e+12  2.236158e+01  9.906869e-01  3.363433e-02  1.728748e+00   
min    -1.534799e+13  0.000000e+00  0.000000e+00 -4.986305e-01 -1.500000e+01   
25%     2.900000e-01  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%     1.300000e+06  0.000000e+00  1.000000e+00  5.480051e-03  0.000000e+00   
75%     3.680000e+07  0.000000e+00  1.000000e+00  2.465802e-02  0.000000e+00   
max     5.400011e+15  1.790700e+04  9.000000e+00  4.986310e-01  1.500000e+01   

                dcml  
count   5.251128e+06  
unique           NaN  
top              NaN  
freq             NaN  
mean    3.269322e+03  
std     9.824728e+03  
min    -1.200000e+01  
25%    -5.000000e+00  
50%    -3.000000e+00  
75%     0.000000e+00  
max     3.276700e+04